### DISCUSSION TITLE
What about middlewares?

### DISCUSSION THREAD

**Discussion by lasseeee** [spatie/laravel-package-tools#6] [Q&A]
What about middlewares?

**Comment by freekmurze** [comment]
What about them?

**Comment by lasseeee** [comment]
Pardon for my briefness. Is there a fluent way to register middlewares in a package using these tools? The GitHub docs seems to not mention middlewares at all. I'm currently doing it through the `bootingPackage()` method the old trusted way:

```
    public function bootingPackage()
    {
        $router = $this->app->make(Router::class);
        $router->pushMiddlewareToGroup('web', MyMiddleware::class);
    }
```

Thanks for your work and for the Package Training course, it's really helpful and insightful 👌

**Reply by freekmurze** [reply]
Thanks for clarifying! I'd accept a PR for this, make sure to include updated tests and readme 👍

**Reply by camya** [reply]
@lasseeee 

I know, this is a little bit off-topic, but I tried your suggested method to push a **middleware to a middleware group** in the package-skeleton-laravel. I try to add a middelware to the 'api' group.

This unfortunately does not work for me. The added middleware is never executed. 

Maybe you can give me a hint how to make it working.

```
public function bootingPackage()
{
    app('router')->pushMiddlewareToGroup('api', UpdateJson::class);
}
```

Strangely, if I register an alias (`app('router')->aliasMiddleware('updatejson', UpdateJson::class)`) at the same position and manually add it to the middleware group, everything works like expected. (But I really need to automatically add the middleware instead of manually adding it.)

Thank you for your help

**Reply by lasseeee** [reply]
I might actually believe you have to store the router in a variable, then call the `pushMiddlewareToGroup` method on that variable.

**Reply by camya** [reply]
Thank you for your help. I tried it. Unfortunately it's the same result. It's really strange, because I read many comments that it should be really straight forward to use pushMiddlewareToGroup(). 

Maybe there is already a package by spatie that uses pushMiddlewareToGroup('api', ...) where I can look how @freekmurze  has implemented it.

I will post an update if I found a solution for my problem. :) Thank you again.

**Reply by camya** [reply]
I fortunately found the solution to run the middleware globally or in a middleware group. 

I implemented the **hasMiddleware() feature** for the spatie/package-skeleton-laravel.

You can take a look at the full [PR hasMiddleware() feature](https://github.com/spatie/laravel-package-tools/pull/41) here.

```
$kernel = app(\Illuminate\Contracts\Http\Kernel::class);
$kernel->prependMiddlewareToGroup($middlewareGroup, $middleware);
```

Btw. I found the solution in the barryvdh/laravel-debugbar implementation. :)
